src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -60,29 +60,48 @@ impl HummockSnapshotManager {
 
     pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> Result<()> {
         tracing::info!("Unpin epoch {} for query {:?}", epoch, &query_id);
-        let local_count = async {
+        let min_epoch = async {
+            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
+            // mark as outdated.
             let mut core_guard = self.core.lock().await;
             let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
             if let Some(query_ids) = query_ids {
                 query_ids.remove(query_id);
+                if query_ids.is_empty() && epoch == core_guard.last_pinned {
+                    core_guard.is_outdated = true;
+                }
+            }
+
+            // Check the min epoch, if the epoch has no query running on it, this should be the min
+            // epoch to be unpin.
+            let mut min_epoch = None;
+            if let Some((epoch, query_ids)) = core_guard.epoch_to_query_ids.first_key_value() {
```

Comment:
One minor optimization:

It is possible that previously, epoch 1 has some queries running, epoch 2,3,4,5 has nothing. No one can be unpinned.
Then in the current function call, epoch 1 turns into no query running. So we are able to unpin 1,2,3,4,5 all together instead of just 1. 

But this case may rarely happen.

LGTM!

Reply:
Indeed.